plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py

```diff
@@ -101,18 +101,23 @@ def destroy(self, dbg, sr):
     def attach(self, dbg, configuration):
         log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
 
-        # ZFS automagically attaches a pool to a mountpoint on
-        # create/boot/etc, so we basically do nothing here but find
-        # this mountpoint
-
-        # FIXME: study how pools are mounted on boot and see if we
-        # could do real a/detach
+        pool_name = configuration["zpool"]
+        try:
+            mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
+        except Exception:
+            log.debug('{}: SR.attach: mountpoint for {} not found, import it'.format(dbg, pool_name))
+        else:
+            # SR.attach should be idempotent. So if the mountpoint is already
+            # mounted just return it.
+            return mountpoint
 
+        zfsutils.pool_import(dbg, pool_name)
         return zfsutils.pool_mountpoint(dbg, configuration["zpool"])
 
     def detach(self, dbg, sr):
-        # Nothing to unmount for now.
-        pass
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        # When exporting, all datasets within the pool are unmounted
+        zfsutils.pool_export(dbg, meta["zpool"])
```

Comment:
Oh in fact it won't work I think. If the user pass a zpool mounted on `/tmp/myzpool` if it is not mounted when rebooting where will we remount it :-(.... I think that in the create if the user pass a zpool that is already mounted we should un-mount it and remount it where we want.

Reply:
So I just tested and it seems ok. After rebooting instead of beeing mounted to `/tmp/myzpool` it will be mounted to `/var/run/sr-mount/myzpool` and it will work. So unless I'm missing something related to DB or some values stored somewhere it should finally be fine.